### DISCUSSION TITLE
AbortController is not defined error in cordova 11.1.0

### DISCUSSION THREAD

**Discussion by ghevge** [apache/cordova#400] [Q&A]
I'm trying to use AbortController  to control the timeout on some fetch backend requests, but it seems it doesn't work in cordova.
Any idea why?

This logic works in Chrome and Edge.

I'm using cordova 11.1.0 .

```
const Timeout = (time) => {
	let controller = new AbortController();
	setTimeout(() => controller.abort(), time);
	return controller;
};
```

**Comment by breautek** [comment]
What normally defines `AbortController`? Where does it come from?

The issue is likely within whatever code that defines that object/class.

**Comment by ghevge** [comment]
It is a javascript object: https://developer.mozilla.org/en-US/docs/Web/API/AbortController

I've found this post related to the same error: https://github.com/ipfs/js-ipfs/issues/4019. Not sure if this applies to cordova too.

**Reply by breautek** [reply]
Ok. For browser features, Cordova support is generally reliant on the underlying system webview. Sometimes plugins may polyfill missing features. But generally speaking you'll have to rely on the compatibility matrix, which MDN has near the bottom of that page.

The browser compatibility table shows the `AbortController` is only available on Safari iOS 12.2 and Chrome Webview 66.

So for android, the first thing I'll check is the webview version the device is running. If you're using an Android emulator, they generally have the factory installed webviews which can be rather dated. (API 24 devices have a factory webview version of like Chrome 53). For android, it's usually pretty safe to rely on `Android WebView` column for compatibility information.

Few different ways to check what webview version you're running on Android:

1. Checking the user agent string
2. If USB debugging is enabled, you can look at `chrome://inspect`
3. On device (API 33, other API versions might be slightly different), go to Settings -> Apps -> All Apps -> Android System Webview. Scroll down to the bottom and the version will be printed.

You can also use a Google Play enabled emulator, which is an emulator that closely resembles a retail device. It has Google Play installed, in which you can login to update the software, which will include system webview updates.

For iOS, it's a bit different, `Safari on iOS` usually refers to the Safari browser app, and not the `WKWebView` that apps uses to embed web content, which is exactly what Cordova does. For the **most** part they are comparable, that is features that are available in Safari are usually available in WKWebView counterpart, but that is not always the case (and there are notable exceptions). However, it's usually always safe to say that the feature won't be available on WKWebView earlier than the `Safari on iOS` compatibility notes, so in this case you'll definitely need iOS 12.2 and later, but further testing may be required to determine if the feature is actually available in WKWebView on iOS 12.2. Hopefully that makes sense...

If you're running in an environment earlier than iOS 12.2 or Chrome 66, you may have to use feature detection to fallback to an alternate strategy (which might involve using the older `XMLHttpRequest` api rather than `fetch` so that you can abort it using `xhr.abort()`).

**Reply by ghevge** [reply]
Thanks for the reply! That error makes sense now.

**Comment by ghevge** [comment]
This is some sample code on how to use AbortController  https://javascript.plainenglish.io/how-to-set-api-timeout-with-javascript-fetch-api-2d7c95782fa9
